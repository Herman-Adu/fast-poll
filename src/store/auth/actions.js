export const SIGN_IN = 'auth/SIGN_IN'
export const SIGN_OUT = 'auth/SIGN_OUT'
export const LOADING = 'auth/LOADING'

export const signInAction = ({ uid, isAnonymous, photoURL }) => ({
  type: SIGN_IN,
  payload: {
    uid,
    isAnonymous,
    photoURL,
  }
})

export const signOutAction = () => ({
  type: SIGN_OUT,
  payload: null
})

export const loadingAction = (isLoading) => ({
  type: LOADING,
  payload: {
    isLoading
  }
})

export const signIn = (auth, provider) => (dispatch) => {
  dispatch(loadingAction(true))
  switch (provider) {
    // the auth listener will handle the success cases
    case 'google':
      return auth()
        .signInWithPopup(new auth.GoogleAuthProvider())
        .then(() => {
          dispatch(loadingAction(false))
        })
        .catch(error => {
          // eslint-disable-next-line no-console
          console.error(error);
          // TODO: notify the user of the error
          return error;
        });

    case 'anonymous':
      return auth()
        .signInAnonymously()
        .then(() => {
          dispatch(loadingAction(false))
        })
        .catch(error => {
          // eslint-disable-next-line no-console
          console.error(error);
          // TODO: notify the user of the error
          return error;
        });

    default:
      const reason = 'Invalid provider passed to signIn method';
      // eslint-disable-next-line no-console
      console.error(reason);
      return Promise.reject(reason);
  }
}

export const signOut = (auth) => (dispatch) => {
  return auth().signOut().then(() => {
    dispatch(signOutAction())
  })
}

export const startListeningToAuthChanges = (auth) => (dispatch) => {
  return auth().onAuthStateChanged(user => {
    if (user) {
      // if user exists sign-in!
      dispatch(signInAction(user))
    } else {
      // otherwise sign-out!
      dispatch(signOutAction())
    }
  });
}