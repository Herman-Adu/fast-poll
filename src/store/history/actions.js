export const LOADING = 'history/LOADING';
export const ADD_POLLS = 'history/ADD_POLLS';

export const loadingAction = isLoading => ({
  type: LOADING,
  payload: {
    isLoading,
  },
});

export const addPollsAction = (type, polls) => ({
  type: ADD_POLLS,
  payload: {
    type,
    polls,
  },
});

export const getPollHistory = (firebase, { uid }) => dispatch => {
  const LIMIT = 12;
  const userRef = firebase.users.doc(uid);

  dispatch(loadingAction(true));

  return Promise.all([
    userRef
      .collection('createdPolls')
      .orderBy('date')
      .limit(LIMIT)
      .get(),
    userRef
      .collection('votedOnPolls')
      .orderBy('date')
      .limit(LIMIT)
      .get(),
  ])
    .then(([created, votedOn]) => {
      const createdPolls = [];
      const votedOnPolls = [];

      created.forEach(doc => {
        createdPolls.push(doc.data());
      });
      votedOn.forEach(doc => {
        votedOnPolls.push(doc.data());
      });

      dispatch(loadingAction(false));
      dispatch(addPollsAction('created', createdPolls));
      dispatch(addPollsAction('votedOn', votedOnPolls));

      return [
        created.docs[created.docs.length - 1],
        votedOn.docs[votedOn.docs.length - 1],
      ];
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.error(error);
      // TODO: notify the user of the error
    });
};
