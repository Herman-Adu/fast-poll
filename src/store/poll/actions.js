export const LOADING = 'poll/LOADING'
export const UPDATE_TITLE = 'poll/UPDATE_TITLE'
export const UPDATE_SELECTION = 'poll/UPDATE_SELECTION'
export const UPDATE_VOTED = 'poll/UPDATE_VOTED'

export const loadingAction = (isLoading) => ({
  type: LOADING,
  payload: {
    isLoading
  }
})

export const updateTitleAction = (title) => ({
  type: UPDATE_TITLE,
  payload: {
    title
  }
})

export const updateSelectionAction = (selection) => ({
  type: UPDATE_SELECTION,
  payload: {
    selection
  }
})

export const updateVotedAction = (hasVoted) => ({
  type: UPDATE_VOTED,
  payload: {
    hasVoted
  }
})

export const updateTitle = (title) => (dispatch) => {
  return dispatch(updateTitleAction(title))
}

export const createPoll = (firebase, history, data) => dispatch => {
  const { pollId, options, title } = data

  dispatch(loadingAction(true))

  return firebase.polls
    .doc(pollId)
    .set({
      title,
      id: pollId,
      options: options.map(({ text, id }) => ({ text, optionId: id })),
    })
    .then(() => {
      dispatch(loadingAction(true))

      history.push(`/poll/${pollId}`);
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.error(error);
      // TODO: notify the user of the error
    });
}