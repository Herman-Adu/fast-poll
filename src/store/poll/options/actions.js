export const ADD_OPTION = 'poll/options/ADD_OPTION'
export const REMOVE_OPTION = 'poll/options/REMOVE_OPTION'
export const UPDATE_OPTION = 'poll/options/UPDATE_OPTION'
export const UPDATE_OPTION_ORDER = 'poll/options/UPDATE_OPTION_ORDER'

export const addOptionAction = (id) => ({
  type: ADD_OPTION,
  payload: {
    id,
  }
})

export const removeOptionAction = (id) => ({
  type: REMOVE_OPTION,
  payload: {
    id
  }
})

export const updateOptionAction = (id, data) => ({
  type: UPDATE_OPTION,
  payload: {
    id,
    data,
  }
})

export const updateOptionOrderAction = (order) => ({
  type: UPDATE_OPTION_ORDER,
  payload: {
    order
  }
})

export const addOption = (id) => (dispatch, getState) => {
  const { data } = getState().poll.options
  // remove any options that dont have text when adding a new one
  Object.entries(data).forEach(([id, option]) => {
    if (!option.text) {
      dispatch(removeOptionAction(id))
    } else {
      dispatch(updateOptionAction(id, { editing: false }))
    }
  })

  dispatch(addOptionAction(id))
}

export const updateOptionOrder = (order) => (dispatch) => {
  dispatch(updateOptionOrderAction(order))
}

export const removeOption = (id) => (dispatch) => {
  dispatch(removeOptionAction(id))
}

export const updateOption = (id, data) => (dispatch) => {
  dispatch(updateOptionAction(id, data))
}