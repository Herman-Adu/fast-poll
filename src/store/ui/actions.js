export const TOGGLE_HEADER_NAV = 'ui/TOGGLE_HEADER_NAV'

export const toggleHeaderNavAction = (isOpen) => ({
  type: TOGGLE_HEADER_NAV,
  payload: {
    isOpen
  }
})

export const toggleHeaderNav = (isOpen) => (dispatch) => {
  dispatch(toggleHeaderNavAction(isOpen))
}