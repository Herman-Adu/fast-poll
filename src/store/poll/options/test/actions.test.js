import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import * as actions from '../actions'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe('poll/options bound actions', () => {
  const state = {
    poll: {
      options: {
        order: ['1', '2', '3'],
        data: {
          '1': {
            id: '1',
            text: '1 poll',
            editing: false,
            votes: 0
          },
          '3': {
            id: '3',
            text: '',
            editing: false,
            votes: 0
          },
          '2': {
            id: '2',
            text: '2 poll',
            editing: false,
            votes: 0
          },
        }
      }
    }
  }
  
  it('addOptionremoves any options with no text when adding a new option', () => {
    const store = mockStore(state)
    store.dispatch(actions.addOption('4'))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('updateOptionOrder', () => {
    const store = mockStore(state)
    store.dispatch(actions.updateOptionOrder(['3', '2', '1']))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('removeOptionremoves', () => {
    const store = mockStore(state)
    store.dispatch(actions.removeOption('3'))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('updateOption', () => {
    const store = mockStore(state)
    store.dispatch(actions.updateOption('3' , { editing: true, text: '3 updated' }))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('resetOptions', () => {
    const store = mockStore(state)
    store.dispatch(actions.resetOptions())
    expect(store.getActions()).toMatchSnapshot()
  })
})