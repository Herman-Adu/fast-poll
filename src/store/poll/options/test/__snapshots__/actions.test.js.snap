// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`poll/options bound actions addOptionremoves any options with no text when adding a new option 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Object {
        "editing": false,
      },
      "id": "1",
    },
    "type": "poll/options/UPDATE_OPTION",
  },
  Object {
    "payload": Object {
      "data": Object {
        "editing": false,
      },
      "id": "2",
    },
    "type": "poll/options/UPDATE_OPTION",
  },
  Object {
    "payload": Object {
      "id": "3",
    },
    "type": "poll/options/REMOVE_OPTION",
  },
  Object {
    "payload": Object {
      "id": "4",
    },
    "type": "poll/options/ADD_OPTION",
  },
]
`;

exports[`poll/options bound actions removeOptionremoves 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "3",
    },
    "type": "poll/options/REMOVE_OPTION",
  },
]
`;

exports[`poll/options bound actions resetOptions 1`] = `
Array [
  Object {
    "payload": null,
    "type": "poll/options/RESET_OPTIONS",
  },
]
`;

exports[`poll/options bound actions updateOption 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Object {
        "editing": true,
        "text": "3 updated",
      },
      "id": "3",
    },
    "type": "poll/options/UPDATE_OPTION",
  },
]
`;

exports[`poll/options bound actions updateOptionOrder 1`] = `
Array [
  Object {
    "payload": Object {
      "order": Array [
        "3",
        "2",
        "1",
      ],
    },
    "type": "poll/options/UPDATE_OPTION_ORDER",
  },
]
`;
